### PR TITLE
🌙 Add Ramadan Mode Feature

### PR DESCRIPTION
**Pull request by karimelgazar**
## Overview

This PR introduces a comprehensive **Ramadan Mode** feature that transforms the prayer calendar into a complete Ramadan companion. When enabled, the feature automatically adds separate calendar events for **Suhoor** (pre-dawn meal), **Iftar** (breaking the fast), and **Tarawih** (night prayers) during Ramadan, making it easier for users to plan their entire Ramadan schedule.

### 🎛️ **Ramadan Mode Toggle**
- Clean toggle switch with moon icon in Advanced Options
- Conditional display of duration controls when enabled
- Responsive design with full RTL support for Arabic
- Visual Ramadan mode indicator throughout the UI

### ⏰ **Three Customizable Ramadan Events** 
- **Suhoor Duration**: 0-120 minutes (default: 30 min) - *Before Fajr prayer*
- **Iftar Duration**: 15-60 minutes (default: 30 min) - *After Maghrib prayer*  
- **Tarawih Duration**: 0-180 minutes (default: 60 min) - *After Isha prayer*
- **Smart Duration Handling**: Events with 0 duration are **not added** to the calendar
- Input validation with min/max constraints and real-time feedback

### 📅 **Intelligent Calendar Generation**
- Automatically detects Ramadan dates using Hijri calendar (month 9)
- Creates **Suhoor** events before Fajr prayer (when duration > 0)
- Creates **Iftar** events after Maghrib prayer (always created)
- Creates **Tarawih** events after Isha prayer (when duration > 0)
- Maintains original prayer events alongside Ramadan events
- Full alarm support for all new events
- **Zero-duration events are excluded** from calendar generation

### 🌍 **Complete Multi-language Support**
- Full English and Arabic translations for all Ramadan elements
- Proper RTL layout with directional toggle switches
- Localized event names in calendar output
- Arabic event names: الإفطار (Iftar), التراويح (Tarawih), السحور (Suhoor)

### 📱 **Enhanced User Interface**
- Visual Ramadan mode indicator in Prayer Preview with moon icon
- Duration badges showing extended times for each event
- Conditional event display based on duration settings
- Consistent dark/light theme support across all new components
- Responsive grid layout for duration inputs

### 🛡️ **Enhanced API Reliability**
- Added retry mechanism with exponential backoff (3 attempts max)
- Implemented 15-second timeout using AbortController  
- Enhanced error handling with detailed logging
- Better debugging information for API connection issues
- Maintains existing caching performance

### 🧪 **Comprehensive Testing Suite**
- Tests for all three Ramadan mode API parameters
- Verification of conditional event creation based on duration
- Arabic language testing for all event names
- Edge case testing: 0 duration handling, non-Ramadan dates
- Suhoor event timing validation (before Fajr)
- Backward compatibility verification

### ⚡ **Performance Optimized**
- Maintains existing multi-layer caching strategy
- No performance impact when Ramadan mode is disabled
- Efficient Hijri date detection with minimal overhead
- Smart event creation only during detected Ramadan days

### New Query Parameters
- `ramadanMode=true` - Enable comprehensive Ramadan mode
- `suhoorDuration=30` - Suhoor event duration in minutes (0 = no event)
- `iftarDuration=30` - Iftar event duration in minutes  
- `traweehDuration=60` - Tarawih event duration in minutes (0 = no event)

### Example URLs
```bash
# Full Ramadan mode
/api/prayer-times.ics?address=Cairo&method=5&ramadanMode=true&suhoorDuration=45&iftarDuration=30&traweehDuration=90

# Ramadan mode without Tarawih (0 duration)
/api/prayer-times.ics?address=Cairo&method=5&ramadanMode=true&suhoorDuration=30&iftarDuration=30&traweehDuration=0
```

## 🎯 Smart Duration Handling

**Important Feature**: Events with duration set to `0` are **not added** to the calendar:

- `suhoorDuration=0` → No Suhoor event created
- `traweehDuration=0` → No Tarawih event created  
- `iftarDuration` always creates an event (minimum 15 minutes)

This allows users to customize their Ramadan experience based on personal or regional preferences.

## 🧪 Testing Coverage

All tests pass with comprehensive coverage:
- ✅ All Ramadan parameter handling and validation
- ✅ Conditional event creation based on duration settings
- ✅ **Zero-duration event exclusion verification**
- ✅ Arabic language support for all events
- ✅ Hijri calendar Ramadan detection
- ✅ API retry mechanisms and timeout handling
- ✅ Backward compatibility with existing functionality

## 📱 User Experience Highlights

1. **Intuitive Discovery**: Ramadan mode prominently displayed with clear descriptions
2. **Visual Feedback**: Moon icons and indicators provide consistent visual language
3. **Flexible Customization**: Three separate duration controls for different needs
4. **Smart Defaults**: Reasonable default durations that work for most users
5. **Cultural Sensitivity**: Full Arabic support with proper RTL layout
6. **Accessibility**: Proper contrast ratios and keyboard navigation

## 🔄 Backward Compatibility

- ✅ **100% backward compatible** - all existing functionality preserved
- ✅ Ramadan mode is **completely opt-in** (disabled by default)
- ✅ **Zero breaking changes** to existing API endpoints
- ✅ Existing calendar subscriptions continue working without modification
- ✅ All existing query parameters function identically

## Summary by CodeRabbit

* **New Features**
  * Introduced "Ramadan Mode" with configurable durations for Iftar, Tarawih, and Suhoor events in the calendar and UI.
  * Ramadan-specific events are now displayed in the prayer preview and included in calendar downloads.
  * Added new translation keys and UI text for Ramadan Mode in both English and Arabic.

* **Bug Fixes**
  * Improved language initialization to prevent hydration mismatches during server-side rendering.

* **Enhancements**
  * Added retry logic and request timeouts for fetching prayer times to improve reliability.
  * Updated ICS link generation to dynamically use the current browser location.

* **Tests**
  * Extended test coverage to validate Ramadan Mode event creation and localization in calendar exports.

### CODE REVIEW

File: src/Components/PrayerPreview.tsx [hunk 1]
```diff
@@ -43,31 +55,66 @@ export default function PrayerPreview({ loadingNext, nextPrayer, todayTimings }:
...
 
+  // Check if today is in Ramadan (simplified check - in a real app you'd use the actual Hijri date)
+  // For now, we'll show Ramadan mode if it's enabled, regardless of the actual date
+  const isRamadanToday = ramadanMode;
+
+  // Check if location is set
+  const hasLocation =
+    locationFields.inputMode === 'address'
+      ? !!locationFields.address.trim()
+      : !!(locationFields.latitude && locationFields.longitude);
```
[the remark this answers is not in the document]
Comment (karimelgazar) [reply src/Components/PrayerPreview.tsx:66]: ## Ramadan Detection Logic

The application uses a **simple but effective approach** to determine if a date is during Ramadan:

### Primary Detection Method

The core logic is in the `isRamadan()` function in `src/app/api/prayer-times.ics/route.ts`:

https://github.com/karimelgazar/prayCalendar/blob/8a8013826b80eac4261cf9f3616e2855468de5dd/src/app/api/prayer-times.ics/route.ts#L45:L53

### How It Works

1. **Data Source**: The application fetches prayer time data from the **AlAdhan API** (`https://api.aladhan.com/v1/calendar/`)

2. **Hijri Calendar**: Each day's data includes both Gregorian and **Hijri calendar information**:
   ```typescript
   date: {
     gregorian: { /* Gregorian date info */ },
     hijri: {
       date: string;
       month: { number: number; en: string; ar: string };
       year: string;
       // ... other Hijri calendar data
     }
   }
   ```

3. **Ramadan Check**: The function simply checks if `day.date.hijri.month.number === 9` because **Ramadan is the 9th month in the Islamic (Hijri) calendar**

### Usage in the Application

The Ramadan check is used in the API route when generating calendar events:

```javascript
// Check if current day is in Ramadan for Ramadan mode
const isRamadanDay = ramadanMode && isRamadan(day);

// Later in the code:
if (isRamadanDay) {
  // Create special Ramadan events (Iftar, Tarawih, Suhoor)
  if (name === 'Maghrib' && iftarDuration > 0) {
    // Create Iftar event after Maghrib
  }
  if (name === 'Isha' && traweehDuration > 0) {
    // Create Tarawih event after Isha  
  }
  if (name === 'Fajr' && suhoorDuration > 0) {
    // Create Suhoor event before Fajr
  }
}
```

### Key Points

- **Accurate**: Uses authentic Islamic calendar data from AlAdhan API
- **Automatic**: No manual date configuration needed - the Hijri calendar automatically accounts for moon sighting variations
- **Global**: Works worldwide since AlAdhan provides location-specific prayer times with correct Hijri dates
- **User Control**: The user must enable "Ramadan Mode" for the special events to be created, but the date detection is automatic

This approach ensures that Iftar, Tarawih, and Suhoor events are only created on actual Ramadan days according to the Islamic calendar, without requiring manual date ranges or assumptions about when Ramadan occurs.

File: src/app/api/prayer-times.ics/route.ts [hunk 2]
```diff
@@ -151,6 +186,96 @@ export async function GET(request: NextRequest) {
...
+                  tarawihEvent.createAlarm({ type: ICalAlarmType.audio, triggerBefore: a * 60 });
+                } else if (a < 0) {
+                  tarawihEvent.createAlarm({ type: ICalAlarmType.audio, triggerAfter: Math.abs(a) * 60 });
+                } else {
+                  tarawihEvent.createAlarm({ type: ICalAlarmType.audio, trigger: 0 });
+                }
+              }
+            }
+          }
+        }
```
[the remark this answers is not in the document]
Comment (karimelgazar) [reply src/app/api/prayer-times.ics/route.ts]: Fixed in [this commit](https://github.com/AhmedElywa/prayCalendar/pull/24/commits/ac7aa0140819d47418e2308d8133237b5f0095c2)

File: src/__tests__/app/api/prayer-times.ics/route.test.ts [hunk 3]
```diff
@@ -336,4 +500,146 @@ describe('Prayer Times API', () => {
...
+    });
+
+    const response = await GET(request);
+
+    expect(response.status).toBe(200);
+
+    // Should create Iftar but not Tarawih when duration is 0
+    expect(mockCreateEvent).toHaveBeenCalledWith(expect.objectContaining({ summary: 'Iftar' }));
+    expect(mockCreateEvent).not.toHaveBeenCalledWith(expect.objectContaining({ summary: 'Tarawih' }));
+  });
```
[the remark this answers is not in the document]
Comment (karimelgazar) [reply src/__tests__/app/api/prayer-times.ics/route.test.ts]: Fixed in [thsi commit](https://github.com/AhmedElywa/prayCalendar/pull/24/commits/54ca4d9050f46cd2a286bd3456d4f7bc2d6bc436)

**Comment by AhmedElywa** [comment]
Hey @karimelgazar,
Thank you for your contribution,
To not make the link to long I suggest to change the link parameters to be only Ramadan 
If it have value will be true 
And the value will be the events time
30,60,30 and in our api we use it like 
[iftar,tarawih,suhoor] = Ramadan.split(,)

I hope it's arrive to you 

Thanks again for your help it's amazing feature
